Index: shadow-4.5/src/userdel.c
===================================================================
--- shadow-4.5.orig/src/userdel.c
+++ shadow-4.5/src/userdel.c
@@ -132,6 +132,12 @@ static int remove_mailbox (void);
 static int remove_tcbdir (const char *user_name, uid_t user_id);
 #endif				/* WITH_TCB */
 
+#ifndef EXTRAUSERS_OPT
+#define EXTRAUSERS_OPT 100000
+#endif
+
+static bool use_extrausers = false;
+
 /*
  * usage - display usage message and exit
  */
@@ -149,6 +155,7 @@ static void usage (int status)
 	(void) fputs (_("  -h, --help                    display this help message and exit\n"), usageout);
 	(void) fputs (_("  -r, --remove                  remove home directory and mail spool\n"), usageout);
 	(void) fputs (_("  -R, --root CHROOT_DIR         directory to chroot into\n"), usageout);
+	(void) fputs (_("      --extrausers              Use the extra users database\n"), usageout);
 #ifdef WITH_SELINUX
 	(void) fputs (_("  -Z, --selinux-user            remove any SELinux user mapping for the user\n"), usageout);
 #endif				/* WITH_SELINUX */
@@ -1006,6 +1013,7 @@ int main (int argc, char **argv)
 			{"help",         no_argument,       NULL, 'h'},
 			{"remove",       no_argument,       NULL, 'r'},
 			{"root",         required_argument, NULL, 'R'},
+                        {"extrausers", no_argument,       NULL, EXTRAUSERS_OPT},
 #ifdef WITH_SELINUX
 			{"selinux-user", no_argument,       NULL, 'Z'},
 #endif				/* WITH_SELINUX */
@@ -1019,6 +1027,9 @@ int main (int argc, char **argv)
 #endif				/* !WITH_SELINUX */
 		                         long_options, NULL)) != -1) {
 			switch (c) {
+                        case EXTRAUSERS_OPT:
+                                use_extrausers = true;
+                                break;
 			case 'f':	/* force remove even if not owned by user */
 				fflg = true;
 				break;
@@ -1162,6 +1173,18 @@ int main (int argc, char **argv)
 		}
 	}
 
+        if (use_extrausers) {
+               pw_setdbname (EXTRAUSERS_PASSWD_FILE);
+               spw_setdbname (EXTRAUSERS_SHADOW_FILE);
+               gr_setdbname (EXTRAUSERS_GROUP_FILE);
+               /* TODO expose this information in other tools */
+               sub_uid_setdbname(EXTRAUSERS_SUBUID_FILE);
+               sub_gid_setdbname(EXTRAUSERS_SUBGID_FILE);
+#ifdef SHADOWGRP
+               sgr_setdbname (EXTRAUSERS_SHADOWGROUP_FILE);
+#endif
+        }
+
 	/*
 	 * Do the hard stuff - open the files, create the user entries,
 	 * create the home directory, then close and update the files.
